xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-macro/macroEditor.js

```diff
@@ -358,73 +115,136 @@ define('macroParameterTreeDisplayer', ['jquery', 'l10n!macroEditor'], function($
+        '</div>' +
+      '</div>',
+
+  displayOptionalNodes = function(parametersMap, optionalNodeList) {
+    let output = $(macroParameterGroupOptionalsTemplate);
+    let tabs = output.find('ul[role="tablist"]');
+    let tabTemplate = tabs.children().remove();
+    let tabPanels = output.find('.tab-content');
+    let tabPanelTemplate = tabPanels.children().remove();
+    optionalNodeList.forEach(function(nodeKey, index) {
+      let childNode = parametersMap[nodeKey];
+      let tab = tabTemplate.clone().appendTo(tabs);
+      let tabPanel = tabPanelTemplate.clone().appendTo(tabPanels);
+      // Some of the child nodes might be hidden so we will activate the first visible tab at the end.
+      tab.add(tabPanel).removeClass('active').toggleClass('hidden', !!childNode.hidden);
+      fillNodeTab(parametersMap, childNode, tab.children().first(), tabPanel);
+    });
+    // Activate the first visible tab.
+    let activeTab = tabs.children().not('.hidden').first();
+    let activeTabPanel = tabPanels.children().not('.hidden').first();
+    activeTab.add(activeTabPanel).addClass('active');
+    toggleMacroParameterGroupVisibility(output);
+    // we remove all occurrences of feature title since we already have the tabs name.
+    output.find('.feature-container .panel-heading').remove();
+    return output;
+  },
+
+  displayGroup = function(parametersMap, groupNode) {
+    return displayFeature(parametersMap, groupNode, groupNode.featureOnly);
+  },
+
+  macroFeatureContainerTemplate =
+    '<div class="feature-container panel panel-default" id="feature-{{featureName}}">' +
+      '<div class="panel-heading">' +
+        '<span class="feature-title">{{featureTitle}}</span>' +
+        '<span class="mandatory"></span>' +
       '</div>' +
-    '</li>',
+      '<div class="panel-body"></div>' +
+    '</div>',
 
-  displayMultipleChoiceGroup = function(groupNode) {
-    var output = $(macroParameterGroupTemplate).addClass('multiple-choice');
-    fillNodeTab(groupNode, output.find('.macro-parameter-group-name'), output.find('.macro-parameter-group-members'));
-    toggleMacroParameterGroupVisibility(output);
+  macroFeatureContentTemplate =
+    '<div class="feature-parameter">' +
+      '<div class="feature-choice">' +
+        '<input type="radio" class="feature-radio" id="feature-radio-{{featureName}}-{{parameterId}}"' +
+      ' name="feature-radio-{{featureName}}" value="{{parameterKey}}" />' +
+        '<label class="feature-choice-name" for="feature-radio-{{featureName}}-{{parameterId}}">' +
+          '{{parameterLabel}}' +
+        '</label>' +
+      '</div>' +
+      '<div class="feature-choice-body"></div>' +
+    '</div>',
+
+  displayFeature = function (parametersMap, featureNode, isFeature) {
+    let name = (isFeature) ? featureNode.featureName : featureNode.name;
+    let output = $(macroFeatureContainerTemplate
+        .replaceAll("{{featureName}}", featureNode.id)
+        .replaceAll("{{featureTitle}}", name));
```

Comment:
I don't like this. Those values may need HTML escaping, so I'd rather look up the parent DOM elements and update the inner text, like you do below.